fix(ssa): Keep last stores used in array returned from a function

vezenovm opened the pull request:
# Description

## Problem\*

Resolves #8786

Builds upon #8743 

## Summary\*

In #8743, we switched to mark all aliases of a Call's arguments unknown. Similarly, in this PR we mark all aliases a return termiantor's arguments unknown. This then keeps the store that would otherwise be removed.

## Additional Context

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

aakoshh commented on compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs:
```diff
@@ -652,10 +652,22 @@ impl<'f> PerFunctionContext<'f> {
+                            });
+                        }
+                    }
+                }
```
Maybe this, and the one [here](https://github.com/noir-lang/noir/blob/8580c15e123c222f49dc55a674f08f026242a525/compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs#L527-L536) and [here](https://github.com/noir-lang/noir/blob/8580c15e123c222f49dc55a674f08f026242a525/compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs#L462-L464) could be merged into a helper function that takes a list value IDs to iterate and a lambda that tells what to do with the alias. Perhaps it could be arranged into further helpers to make it more intuitive to the casual reader to connect "anything which acts on a reference in these values" to gathering aliases.

vezenovm replied:
> It doesn't look like this handles the cases where expressions or aliases may return `None` - e.g. expressions with unknown aliases.

@jfecher I added that test. `v0` will still be included as an alias of v4, so we will not remove `store u1 1 at v0`.

aakoshh commented:
I think the fuzzer still finds failures after this; I'll try to isolate them in new tickets.

vezenovm commented:
In the parent PR https://github.com/noir-lang/noir/pull/8743 we were not appropriately labeling array aliases. To keep arrays containing references we updated the input to `mark_all_unknown` by collecting all aliases for every argument to a call or return. In fact, `mark_all_unknown` was just too restrictive. I have updated `mark_all_unknown` to access every alias from an array constant.

vezenovm commented:
This is ready for review again. cc @jfecher
